From 342a1a1cf431401d0b61fb22943a30efb8ed8ace Mon Sep 17 00:00:00 2001
From: oli <olit@google.com>
Date: Thu, 27 Nov 2025 16:27:33 +0000
Subject: [PATCH] Add resolveActivityAsUserForExplicitType api to pm

allow checking for resolved activity for an intent with a specific type

allowing the intent to declare its own type can lead to a security
vulnerability if the intent changes its type after the IntentForwarder#canForward
check.

Use the new API in IntentForwarderActivity to prevent checking the intent's type twice.

Update other call sites of IntentForwarder#canForward to extract the resolved type from the intent and pass in to to the method call

Check the launch package rather than calling package

sanitize intent selector as well as received intent

Signed-off-by: oli <olit@google.com>
---
 .../internal/app/AbstractMultiProfilePagerAdapter.java        | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/java/com/android/internal/app/AbstractMultiProfilePagerAdapter.java b/core/java/com/android/internal/app/AbstractMultiProfilePagerAdapter.java
index 23167382ec3b..6e534ce3c7b9 100644
--- a/core/java/com/android/internal/app/AbstractMultiProfilePagerAdapter.java
+++ b/core/java/com/android/internal/app/AbstractMultiProfilePagerAdapter.java
@@ -351,7 +351,7 @@ public abstract class AbstractMultiProfilePagerAdapter extends PagerAdapter {
 
             return intents.stream().anyMatch(intent ->
                     null != IntentForwarderActivity.canForward(intent, source, target,
-                            packageManager, mContentResolver));
+                            packageManager, intent.resolveTypeIfNeeded(mContentResolver)));
         }
     }
 
@@ -592,4 +592,4 @@ public abstract class AbstractMultiProfilePagerAdapter extends PagerAdapter {
          */
         boolean isWaitingToEnableWorkProfile();
     }
-}
\ No newline at end of file
+}
-- 
2.34.1

